### PR TITLE
Document updates

### PR DESCRIPTION
**Pull request by colindresj**
This includes a bunch of updates to how you use documents. The AST is now created by https://github.com/tendant/graphql-clj, which means we can use some of the features made available by https://github.com/retro/graphql-builder.

A few wins here:

- No need to add `__typename` everywhere -- this fields gets automatically added on parse
- Fragments are now supported within queries
- You can do run-time composition of parsed documents

So, while documents still remain static, you can now bring multiple parsed documents together at run-time using the `compose` function:

```clj
(ns my-app.queries)

(def me-query
  (artemis.document/parse-document
    "query FetchMe {
      me {
        id
        ...PersonName
      }
    }"))

(ns my-app.a.fragments)

(def person-name
  (artemis.document/parse-document
    "fragment PersonName on Person {
      firstName
      lastName
    }"))

(ns my-app.a.core
  (:require [my-app.a.queries :refer [me-query]]
            [my-app.a.fragments :refer [person-fields]]))

(artemis.document/compose me-query person-fields)
```
Furthermore, you may want to use the same query for multiple resources, so you can attach a mapping to a composed document, so that variables can be resolved for each operation:

```clj
(ns my-app.queries)

(def person-query
  (artemis.document/parse-document
    "query Person($slug: ID!) {
      person(slug: $slug) {
        id
        ...PersonName
      }
    }"))

(ns my-app.a.core
  (:require [my-app.a.queries :refer [me-query]]
            [my-app.a.fragments :refer [person-fields]]))

(def get-a-and-b
  (-> person-query
      (artemis.document/compose person-fields)
      (artemis.document/with-mapping {:person-a "Person" :person-b "Person"}))

;; This will send over a query composed of two "Person" query operations, each with the correct variables applied
(artemis.core/query! get-a-and-b {:person-a {:slug "person-a"} :person-b {:slug "person-b"}})
```

Decent amount of code here, so happy to add more comments or make changes.

### CODE REVIEW

**Comment by yedi** [comment]
sweet, will take a closer look sometime today

File: src/artemis/document.cljc [hunk 1]
```diff
@@ -1,66 +1,123 @@
...
+  Different keys can map to the same operation. When running an operation from
+  the document, the operation mapping will be used to resolve multiple sets of
+  variables based on the mapped key."
   {:added "0.1.0"}
-  [document]
-  (:ast document))
+  [doc mapping]
+  (vary-meta doc assoc ::operation-mapping mapping))
+
+#?(:clj (do
```
Comment (dkozma) [inline src/artemis/document.cljc]: Indentation is weird here - what is the `do` for?

Comment (colindresj) [reply src/artemis/document.cljc]: When you're doing this reader macro stuff, you can only give it a single expression, so if you want to do a bunch of stuff (like define things) you have to wrap them in a do:

Check out the section right Host Interop here https://clojure.org/guides/reader_conditionals
As for indentation, that's just the style I've seen most often, bringing that `(do` part up there 
and then pretending it doesn' exist. I'm happy to follow the "rules" a bit more with keeping the expressions in line with the `do`

Comment (dkozma) [reply src/artemis/document.cljc]: That's cool, I'm just unfamiliar with this syntax/style - thanks for the info.

**Comment by yedi** [comment]
🍔 good stuff

**Pull request merged**